crash_rate_aggregates/fill_database.py

```diff
@@ -0,0 +1,192 @@
+    "environment/settings/e10sEnabled",
+]
+
+# names of the comparable dimensions above, used as dimension names in the database
+DIMENSION_NAMES = [
+    "build_version",
+    "build_date",
+    "channel",
+    "application",
+    "os_name",
+    "os_version",
+    "architecture",
+    "country",
+    "experiment_id",
+    "experiment_branch",
+    "e10s_enabled",
+]
+assert len(COMPARABLE_DIMENSIONS) == len(DIMENSION_NAMES)
+
+INSERT_CHUNK_SIZE = 500 # number of records to accumulate in a single database request; higher values mean faster database insertion at the expense of memory usage
+
+def compare_crashes(pings, start_date, end_date, comparable_dimensions, dimension_names):
+    """Returns a PairRDD where keys are user configurations and values are Numpy arrays of the form [usage hours, main process crashes, content process crashes, plugin crashes]"""
+    ping_properties = get_pings_properties(pings, comparable_dimensions + [
+        "meta/submissionDate",
+        "creationDate",
+        "payload/info/subsessionLength",
+        "meta/docType",
+        "payload/keyedHistograms/SUBPROCESS_CRASHES_WITH_DUMP/content",
+        "payload/keyedHistograms/SUBPROCESS_CRASHES_WITH_DUMP/plugin",
+        "payload/keyedHistograms/SUBPROCESS_CRASHES_WITH_DUMP/gmplugin",
+    ], with_processes=True)
+    def is_valid(ping): # sanity check to make sure the ping is actually usable for our purposes
+        submission_date = ping["meta/submissionDate"]
+        if not isinstance(submission_date, str) and not isinstance(submission_date, unicode):
+            return False
+        activity_date = ping["creationDate"]
+        if not isinstance(activity_date, str) and not isinstance(activity_date, unicode):
+            return False
+        subsession_length = ping["payload/info/subsessionLength"]
+        if subsession_length is not None and not isinstance(subsession_length, int) and not isinstance(subsession_length, long):
+            return False
+        return True
+
+    def get_crash_pair(ping): # responsible for normalizing a single ping into a crash pair
+        # we need to parse and normalize the dates here rather than at the aggregates level,
+        # because we need to normalize and get rid of the time portion
+
+        # date the ping was received
+        submission_date = datetime.strptime(ping["meta/submissionDate"], "%Y%m%d").date() # convert the YYYYMMDD format to a real date
+        submission_date = max(start_date, min(end_date, submission_date)) # normalize the submission date if it's out of range
+
+        # date the ping was created on the client
+        try:
+            activity_date = dateutil.parser.parse(ping["creationDate"]).date() # the activity date is the date portion of creationDate
+        except: # malformed date, like how sometimes seconds that are not in [0, 59]
+            activity_date = submission_date # we'll just set it to the submission date, since we can't parse the activity date
```

Comment:
I would prefer to ignore invalid data, as long as we count how many pings have been ignored.